### PR TITLE
Add Phi4 INT4 corpus mismatch diagnostics

### PR DESCRIPTION
**Pull request by DeanoC**
## Summary
- extend Phi4 debug logits dumping to retain top-k samples for the sampled prefill/decode steps
- add corpus-comparer mismatch diagnostics that capture Python top-k logits, rerun Rust with materialized logits, and classify near-tie divergences
- include diagnostic details in the JSON report for follow-up on direct CUDA argmax disagreements

## Validation
- `SUPERSONIC_BACKENDS=cuda cargo build --release --bin supersonic`
- `python3 -m py_compile oracle/int4_corpus_compare.py`
- `rustfmt --check crates/runner/src/phi4_engine.rs`
- `git diff --check`
- focused diagnostic run: `HF_HOME=/dev/shm/hf_home SUPERSONIC_BACKENDS=cuda python3 oracle/int4_corpus_compare.py --model-dir /dev/shm/Phi-4-mini-int4-oracle --model-variant phi4-mini --bake-subdir .supersonic/v2-int4-gptq --max-new-tokens 8 --prompts /tmp/phi4_focus_prompts.json --binary target/release/supersonic --device cuda:0 --diagnose-mismatches --report /tmp/phi4_focus_diagnostics_report.json` (expected exit 1: 2/3 prompts matched; divergent moon prompt classified as near_tie=True with exact 0.0 BF16 gaps)

### CODE REVIEW

**Comment by DeanoC** [comment]
Addressed both review comments in 218ee0a:

- top-level Phi4 debug logits now keep `step`, `next`, and `top` tied to the current sampled step, while `samples` retains the full history.
- mismatch diagnostics now check that the materialized Rust rerun matches the original Rust token prefix before the divergence. If it does not, the report sets `prefix_aligned=false` and skips near-tie gap classification instead of using logits from a different context.

Revalidated with build, py_compile, rustfmt --check, git diff --check, and the focused diagnostic corpus run.

**Comment by DeanoC** [comment]
@codex review

**Comment by DeanoC** [comment]
Addressed the latest review comment in a87e073:

- `topk_payload` now returns an empty payload for non-positive top-k values and clamps oversized values to the logits length before calling `torch.topk`.
- The corpus loop now only asks Python to collect top-k payloads when `--diagnose-mismatches` is enabled, so normal comparison runs do not pay diagnostic overhead and cannot fail from `--diagnostic-top-k`.

Validation:
- `python3 -m py_compile oracle/int4_corpus_compare.py`
- `git diff --check`
- oversized-top-k smoke run with `--diagnostic-top-k 9999999` and diagnostics disabled: 1/1 prompt matched.
